[Lee Kang Quan] ip

Kang-Quan opened the pull request:
### Duke

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duke frees your mind of having to remember things you need to do. It is,

- text-based
- easy to learn
- ~beautiful~ very beautiful user interface

**What are you waiting for??**
1. download it from [here](https://github.com/Kang-Quan/ip/releases/download/A-Jar/duke.jar)
2. double-click it
3. add your tasks
4. let it remember them for you :wink:

And it is **FREE**!!!

Features:

- [x] Managing Tasks
- [x] Managing deadlines
- [ ] Reminders (COMING SOON)<hr/>

If you Java programmer, you can use it to practice Java too. Here's the <span style="color:red">`main`</span> method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Kang-Quan committed:
Add echo and bye to duke

Kang-Quan committed:
Add list and add to duke

Kang-Quan committed:
Add mark as done and undo mark as done to duke

Kang-Quan committed:
Add ToDos, Events and Deadlines to duke

Kang-Quan committed:
Add TextUiTesting to duke

Kang-Quan committed:
Add handle error message to duke

Kang-Quan committed:
Add delete function to duke

Kang-Quan committed:
Add enum to replace keyword

Kang-Quan committed:
Add save function to duke

Kang-Quan committed:
Add date, time to event and deadline tasks

Kang-Quan committed:
Merge branch 'branch-Level-7'

Kang-Quan committed:
Merge branch 'branch-Level-8'

Kang-Quan committed:
Fix bug in loading SavedTask.txt to list

Kang-Quan committed:
Refactor code into OOP

Kang-Quan committed:
Add package to duke

jorrdansoh commented on src/main/java/command/ByeCommand.java:
```diff
@@ -0,0 +1,23 @@
+    @Override
+    public boolean isExit() {
+        return true;
+    }
```
Perhaps you can write some JavaDoc comments for public methods/classes? I noticed the same issue in several other places too.

jorrdansoh commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,20 @@
+
+    public String toSave() {
+        return String.format("[D]" + "[%s] " + super.toString() + " (by: " + super.getDateSave() + ")", super.getStatusIcon());
+    }
```
Since the return statements are quite long, perhaps you can use line wrapping to make them more readable? I noticed the same issue in other places too.

jorrdansoh commented on src/main/java/utility/Parser.java:
```diff
@@ -0,0 +1,158 @@
+            }
+            return new ListCommand();
+
+            case "mark":
```
For `switch` statements, the `case`s should be at the same indentation level as`switch` :)

I noticed this issue in the `Storage` and `Ui` classes too.

jorrdansoh commented on src/main/java/utility/Parser.java:
```diff
@@ -0,0 +1,158 @@
+                if (!validTime) {
+                    throw new DukeException("invalid command deadline");
+                }   
+                return new TimeTaskCommand(input, "deadline", LocalDateTime.parse(by, formatter));
```
The `switch` statement is rather long, perhaps you should consider extracting some of the logic into the respective classes/new classes?

jorrdansoh commented on src/main/java/utility/Ui.java:
```diff
@@ -0,0 +1,133 @@
+import java.util.Scanner;
+
+public class Ui {
+    private static final String DIVIDER = "\t___________________________\n";
```
I like how you have a static variable for the divider to reduce code duplication :+1:

jorrdansoh commented on src/main/java/utility/Storage.java:
```diff
@@ -0,0 +1,132 @@
+                    desciption = input.substring(8, input.lastIndexOf(": ") - 4);
+                    formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
+                    date = LocalDateTime.parse(at, formatter);
+                    newTask = new Deadline(desciption, date);
```
I noticed that the code to get a `newTask` is the same for `case`s `D` and `E`, perhaps you can extract the code out to a method to reduce boilerplate?

jorrdansoh reviewed (commented):
Overall, I found your code easy to read for the most part except a few places
where the methods were too long. I noted a few minor coding standard violations
too. Some of the methods/classes are getting quite long. Consider splitting into
smaller classes if that makes sense.

Kang-Quan committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

Kang-Quan committed:
Add gradle to duke

Kang-Quan committed:
Add junit test to class TestList

Kang-Quan committed:
Add javadoc to duke

Kang-Quan committed:
Refactor code according to coding standard

Kang-Quan committed:
Add find function to duke

Kang-Quan committed:
Merge branch 'branch-A-CodingStandard'

Kang-Quan committed:
Merge branch 'branch-A-JavaDoc'

Kang-Quan committed:
Merge branch 'branch-Level-9'

garfield-oo7 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,41 @@
+
+   
+    public void run() {
+        ui.showWelcome();
```
Would it be better to name the method displayWelcomeText() or greetUser()?

garfield-oo7 commented on src/main/java/utility/Ui.java:
```diff
@@ -0,0 +1,196 @@
+
+        case "invalid command unmark":
+        printWithDivider("\t☹ OOPS!!! It must be in the format of: unmark <position in list>\n");
+        break;
+
+        case "empty command delete":
+        printWithDivider("\t☹ OOPS!!! The description of delete cannot be empty.\n");
+        break;
+
+        case "invalid command delete":
+        printWithDivider("\t☹ OOPS!!! It must be in the format of: delete <position in list>\n");
+        break;
+
+        case "empty todo":
+        printWithDivider("\t☹ OOPS!!! The description of todo cannot be empty.\n");
+        break;
+
+        case "empty deadline":
+        printWithDivider("\t☹ OOPS!!! The description of deadline cannot be empty.\n");
+        break;
+
+        case "invalid command deadline":
+        printWithDivider("\t☹ OOPS!!! It must be in the format of: deadline <desciption> /by <yyyy-mm-dd HH:MM>\n");
+        break;
+
+        case "empty event":
+        printWithDivider("\t☹ OOPS!!! The description of event cannot be empty.\n");
+        break;
+
+        case "invalid command event":
+        printWithDivider("\t☹ OOPS!!! It must be in the format of: event <desciption> /at <yyyy-mm-dd HH:MM>\n");
+        break;
+
+        case "empty taskslist":
+        printWithDivider("\tTasklist is empty\n");
+        break;
+
+        case "empty command find":
+        printWithDivider("\t☹ OOPS!!! The description of find cannot be empty.\n");
+        break;
```
I like how you are following the coding guidelines!

garfield-oo7 commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,33 @@
+ */
+public class Deadline extends TimeTask {
+
+    private final static String ICON = "D";
```
Great use of static variable 👍 !

garfield-oo7 commented on src/main/java/utility/Ui.java:
```diff
@@ -0,0 +1,196 @@
+     * Prints welcome text
+     */
+    public void showWelcome() {
+        String logo = " ____        _        \n"
+                + "|  _ \\ _   _| | _____ \n"
+                + "| | | | | | | |/ / _ \\\n"
+                + "| |_| | |_| |   <  __/\n"
+                + "|____/ \\__,_|_|\\_\\___|\n";
```
Would it be better to instead use this a static variable?

garfield-oo7 reviewed (commented):
Overall, the code was very easy to follow and I like how you have followed the coding standards very closely.

garfield-oo7 commented on src/main/java/command/FindCommand.java:
```diff
@@ -0,0 +1,20 @@
+        this.keyword = keyword;
+    }
+
+
```
Would it be possible to add javadocs?

Piyotato commented on src/main/java/utility/Storage.java:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    /**
+     * Instantiates a new mark command
```
I think this comment is incorrect

Piyotato commented on src/main/java/command/ListCommand.java:
```diff
@@ -0,0 +1,30 @@
+import utility.Ui;
+
+/**
+ * Represents command for list keyword
```
I think this can be more descriptive - how about "Represents command for enumerating task list"?

Piyotato commented on src/main/java/command/ListCommand.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    /**
+     * Executes the list command
```
Same comment as above - this doesn't really explain what the list command actually does.

Piyotato commented on src/main/java/command/Command.java:
```diff
@@ -0,0 +1,45 @@
+     */
+    public abstract void execute(TaskList tasks, Ui ui, Storage storage);
+
+    /**
+     * Returns true if bye command is executed to
+     * end the program
+     * @return Returns true if bye command is executed
+     */
+    public boolean isExit() {
+        return false;
+    }
```
Should this really be part of `command`?

Piyotato commented on src/main/java/command/Command.java:
```diff
@@ -0,0 +1,45 @@
+public abstract class Command {
+    private String input;
+
+    /**
+     * Instantiates a new command
+     */
```
I think this is not very descriptive. Also, I think you forgot to include the parameters in this JavaDoc.

Piyotato commented on src/main/java/command/DeleteCommand.java:
```diff
@@ -0,0 +1,40 @@
+     * @param storage To write and read from a text file
+     */
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+        //create new task here
```
I think this is a misplaced comment?

Piyotato commented on src/main/java/task/TimeTask.java:
```diff
@@ -0,0 +1,40 @@
+ */
+public class TimeTask extends Task {
+
+    public LocalDateTime date;
```
Does this need to be public?

Piyotato commented on src/main/java/tasklist/TaskList.java:
```diff
@@ -0,0 +1,142 @@
+     * Instantiates a new TaskList with empty list
+     */
+    public TaskList() {
+        this.tasks = new ArrayList<Task>(100);
```
I don't think there is a need to maintain the 100 size limit after changing to an 'ArrayList'

Piyotato commented on src/main/java/tasklist/TaskList.java:
```diff
@@ -0,0 +1,142 @@
+    }
+
+
+    /**
```
There seems to be quite some duplicated code here. I think there is some abstraction to be made to combine these two methods.

Piyotato commented on src/main/java/utility/Storage.java:
```diff
@@ -0,0 +1,150 @@
+     *
+     * @param taskList the list we add into the text file
+     */
+    public void update(List<Task> taskList) {
```
I think the method name can be more descriptive.

Kang-Quan committed:
Add GUI to duke

Kang-Quan committed:
Merge branch 'branch-Level-10'

Kang-Quan committed:
Add Assertions to Duke

Assertions allow for easy debugging when we plan to extend our codebase
in the future iterations

Assertions are added to constructors to make sure Tasks object and
String objects are not null.

Assertions are added to TaskList methods to ensure all index are
non-negative and smaller than size of taskList.

Kang-Quan committed:
Refactor folder named Duke to duke

name representing package should be in lowercase according to
java standard

Let's pull up all file that use package Duke and rename it to
duke.

Change all file that uses Duke in its file path and rename it
to duke

Kang-Quan committed:
Refactor codebase to match coding standard

Remove dead code in Ui.java where printDivider method is not used.

Refactor javadocs to meet coding standard and update javadocs for
methods in Ui because they returns a string now instead of void.

Kang-Quan committed:
Refactor code to into shorter method

Refactor code in Parser, Storage and Ui
and shorten long methods to less than 30 LOC.

Kang-Quan committed:
Merge pull request #2 from Kang-Quan/branch-A-CodeQuality

Improve code quality of code base

Kang-Quan committed:
Merge branch 'master' into branch-A-Assertions

Resolve merge conflict when merging

Kang-Quan committed:
Merge pull request #3 from Kang-Quan/branch-A-Assertions

Add Assertions to Duke

Kang-Quan committed:
Add feature to prevent duplicate tasks

Kang-Quan committed:
Merge pull request #4 from Kang-Quan/branch-BCD-Extension

Add feature to prevent duplicate tasks

Kang-Quan committed:
Add greetings and help command to duke

Kang-Quan committed:
Merge pull request #5 from Kang-Quan/branch-Greetings

Add greetings and help command to duke

Kang-Quan committed:
Add QuickStart image for user guide

Kang-Quan committed:
Add product screenshot

Kang-Quan committed:
Delete QuickStart.png

Kang-Quan committed:
Add QuickStart.png for user guide

Kang-Quan committed:
Delete QuickStart.png

Kang-Quan committed:
Add QuickStart.png for user guide

Kang-Quan committed:
Delete QuickStart.png

Kang-Quan committed:
Add QuickStartImg.png for user guide

Kang-Quan committed:
Add junit test and add new image and add javadocs

Kang-Quan committed:
Update README.md into user guide

Kang-Quan committed:
Update README.md

Kang-Quan committed:
Update README.md

Kang-Quan committed:
Update build.gradle

Kang-Quan committed:
Update README.md

Kang-Quan committed:
Update README.md

Kang-Quan committed:
Fix file path of SavedTask.txt to relative path

Kang-Quan committed:
Merge branch 'master' of https://github.com/Kang-Quan/ip

Kang-Quan committed:
Fix bug to create txt file if it does not exist

Kang-Quan committed:
Remove unused code from codebase

Kang-Quan committed:
Update README.md file into user guide

Kang-Quan committed:
Refactor code to make it shorter

Kang-Quan committed:
Add javadocs to method in Storage.java

Kang-Quan committed:
Update README.md

Kang-Quan committed:
Update README.md
